### PR TITLE
feat: extend TripBloc with trip history functions

### PR DESCRIPTION
**Pull request by Nick-Zag**

### CODE REVIEW

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 1]
```diff
@@ -178,4 +215,54 @@ export class TripBloc {
...
+    this.upcomingTrips$.next(upcomingTrips)
+  }
+
+  /**
+   * Requests completed trips list
+   */
+  async searchPastTrips() {
+    this.pastTripsRequests++
+    const paginationOffset = this.pastTripsRequests === 1 ? 0 : this.pastTripsAmount
+    const params = {
```
Comment (andrei-oksentyk) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:243]: Looks like a "dirty hack". Could you explain your solution?

File: packages/demand-bloc-track-trip/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+export const makeSearchParams = (
+  statuses: TripStatuses[],
+  paginationOptions?: Pick<SearchParams, 'pagination_offset' | 'pagination_row_count'>
+): Partial<SearchParams> => {
+  return {
+    trip_states: statuses,
+    pagination_offset: paginationOptions?.pagination_offset || 0,
+    pagination_row_count: paginationOptions?.pagination_row_count || 2,
+    order_by: ['date'],
+  }
```
Comment (andrei-oksentyk) [inline packages/demand-bloc-track-trip/src/utils.ts:13]: `order_by` could contain other options of sorting (see [swagger](https://github.com/karhoo/api-v1-bookings/blob/818adaf772927457cbecf6ba8ccee839cc367dc5/docs/swagger/bookings.yaml#L1752)). I'd propose to define enum for it

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 3]
```diff
@@ -40,18 +57,35 @@ export class TripBloc {
...
     this.storage = storage
+    this.tripsOffset = {
+      upcomingTripsOffset: 0,
+      pastTripsOffset: 0,
+    }
+    this.options = options || {
+      paginationRowCount: {
+        upcomingTrips: defaultUpcomingTripsRow,
+        pastTrips: defaultPastTripsRow,
+      },
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:87]: From my perspective, it's better to keep a single key not object

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 4]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
...
+  ) {
+    const params = makeSearchParams(statuses, paginationOptions)
+
+    const response = await this.tripService.search(params)
+
+    if (!response.ok) return
+
+    const trips: Array<BookATripResponse> = response.body?.bookings || []
+
+    return trips
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:249]: `return response.body?.bookings || []`

Comment (juliamaksimchik) [reply packages/demand-bloc-track-trip/src/TripBloc.ts:249]:
```
async searchTrips(
    statuses: TripStatuses[] = [],
    paginationOptions?: Pick<SearchParams, 'pagination_offset' | 'pagination_row_count'>
  ): Array<BookATripResponse>
```

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 5]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
     this.trackSubscription.unsubscribe()
     this.fareSubscription.unsubscribe()
   }
+
+  /**
+   * Requests trips list
+   * @param statuses means trips with specified statuses
+   * @param paginationOptions for handle pagination
+   */
+  async searchTrips(
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:237]: should it be private ?

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 6]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
...
+    paginationOffset: number
+  ) {
+    if (paginationOffset === 0) {
+      this.tripsOffset[tripsPagination] = 0
+    }
+
+    const statuses =
+      stream === this.upcomingTrips$
+        ? [TripStatuses.CONFIRMED, TripStatuses.REQUESTED]
+        : [TripStatuses.COMPLETED]
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:269]: ugh, don't really like this condition

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 7]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
...
+    const trips = await this.searchTrips(statuses, {
+      pagination_offset: paginationOffset,
+    })
+
+    this.tripsOffset[tripsPagination] += trips?.length || 0
+    stream.next(trips)
+  }
+
+  getUpcomingTrips(paginationOffset = 0) {
+    console.log('GET UPCOMING TRIPS BLOC')
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:280]: pls remove

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 8]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
...
+    return this.getTrips(this.upcomingTrips$, 'upcomingTripsOffset', paginationOffset)
+  }
+
+  getNextUpcomingTrips() {
+    return this.getUpcomingTrips(this.tripsOffset.upcomingTripsOffset)
+  }
+
+  async getPastTrips(paginationOffset = 0) {
+    console.log('GET PAST TRIPS BLOC')
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:289]: pls remove

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 9]
```diff
@@ -178,4 +228,81 @@ export class TripBloc {
...
+    return this.getTrips(this.pastTrips$, 'pastTripsOffset', paginationOffset)
+  }
+
+  getNextPastTrips() {
+    console.log('GET NEXT PAST TRIPS BLOC')
```
Comment (juliamaksimchik) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:294]: pls remove

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 10]
```diff
@@ -5,16 +5,21 @@ import {
...
 import { publishReplay, refCount } from 'rxjs/operators'
-import { FinalTripStatuses, FinalFareStatuses } from './statuses'
+import { FinalTripStatuses, FinalFareStatuses, TripStatuses } from './statuses'
 import { tripTransformer, TripFollowResponse } from './tripTransformer'
+import { makeSearchParams } from './utils'
 
 const POLLING_INTERVAL_TRACK = 10000
 const POLLING_FINAL_FARE = 20000
+const defaultUpcomingTripsRow = 2
+const defaultPastTripsRow = 10
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:22]: I'd suggest to stick to the same code styling for constants, either `THIS_STYLE` or `thatStyle`

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 11]
```diff
@@ -25,13 +30,25 @@ export interface Storage {
...
 }
 
-export type TripService = Pick<Trip, 'cancelByFollowCode' | 'trackTrip'>
+export type TripService = Pick<Trip, 'cancelByFollowCode' | 'trackTrip' | 'search' | 'cancel'>
 export type FareService = Pick<Fare, 'status'>
+export type TripsOffset = {
+  upcomingTripsOffset: number
+  pastTripsOffset: number
+}
+export type CustomOptions = {
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:39]: Add an empty line before this line please

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 12]
```diff
@@ -40,18 +57,35 @@ export class TripBloc {
...
+  ) {
     this.tripService = tripService
     this.fareService = fareService
 
     this.storage = storage
+    this.tripsOffset = {
+      upcomingTripsOffset: 0,
+      pastTripsOffset: 0,
+    }
+    this.options = options || {
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:83]: Add an empty line before this line please

File: packages/demand-bloc-track-trip/src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+export const makeSearchParams = (
+  statuses: TripStatuses[],
+  paginationOptions?: Pick<SearchParams, 'pagination_offset' | 'pagination_row_count'>
+): Partial<SearchParams> => {
+  return {
+    trip_states: statuses,
+    pagination_offset: paginationOptions?.pagination_offset || 0,
+    pagination_row_count: paginationOptions?.pagination_row_count || 2,
+    order_by: ['date'],
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/utils.ts:12]: Might be good to create a constant for `'date'`

File: packages/demand-bloc-track-trip/src/TripBloc.ts [hunk 14]
```diff
@@ -25,13 +30,25 @@ export interface Storage {
...
+export type TripsOffset = {
+  upcomingTripsOffset: number
+  pastTripsOffset: number
+}
+export type CustomOptions = {
+  paginationRowCount: {
+    upcomingTrips: number
+    pastTrips: number
+  }
+}
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/TripBloc.ts:44]: Is nesting really necessary here? I'd avoid nesting unless there are strong reasons for it.

File: packages/demand-bloc-track-trip/src/utils.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  paginationOptions?: Pick<SearchParams, 'pagination_offset' | 'pagination_row_count'>,
+  order?: OrderOptions[]
+): Partial<SearchParams> => {
+  return {
+    trip_states: statuses,
+    pagination_offset: paginationOptions?.pagination_offset || 0,
+    pagination_row_count: paginationOptions?.pagination_row_count || 2,
+    order_by: order || ['date'],
+  }
+}
```
Comment (aliaksandr-bordak) [inline packages/demand-bloc-track-trip/src/utils.ts:28]: `return` is redundant here
```
... => ({
    trip_states: statuses,
    pagination_offset: paginationOptions?.pagination_offset || 0,
    pagination_row_count: paginationOptions?.pagination_row_count || 2,
    order_by: order || ['date'],
})
```

**Review by juliamaksimchik — APPROVED** [review]

**Review by andrei-oksentyk — APPROVED** [review]

**Review by aliaksandr-bordak — APPROVED** [review]

**Review by aliaksandr-bordak — CHANGES_REQUESTED** [review]
Rebase this branch because there are 3 tags missing: 15.1, 15.2, 15.3

**Review by aliaksandr-bordak — APPROVED** [review]

**Pull request merged**